planner/core/rule_partition_processor.go

```diff
@@ -742,28 +742,42 @@ func makePartitionByFnCol(sctx sessionctx.Context, columns []*expression.Column,
 	var fn *expression.ScalarFunction
 	switch raw := partExpr.(type) {
 	case *expression.ScalarFunction:
-		// Special handle for floor(unix_timestamp(ts)) as partition expression.
-		// This pattern is so common for timestamp(3) column as partition expression that it deserve an optimization.
-		if raw.FuncName.L == ast.Floor {
-			if ut, ok := raw.GetArgs()[0].(*expression.ScalarFunction); ok && ut.FuncName.L == ast.UnixTimestamp {
-				args := ut.GetArgs()
-				if len(args) == 1 {
-					if c, ok1 := args[0].(*expression.Column); ok1 {
+		args := raw.GetArgs()
+		switch raw.FuncName.L {
+		case ast.Floor:
+			// Special handle for floor(unix_timestamp(ts)) as partition expression.
+			// This pattern is so common for timestamp(3) column as partition expression that it deserve an optimization.
+			if ut, ok := args[0].(*expression.ScalarFunction); ok && ut.FuncName.L == ast.UnixTimestamp {
+				args1 := ut.GetArgs()
+				if len(args1) == 1 {
+					if c, ok1 := args1[0].(*expression.Column); ok1 {
 						return c, raw, monotoneModeNonStrict, nil
 					}
 				}
 			}
-		}
-
-		fn = raw
-		args := fn.GetArgs()
-		if len(args) > 0 {
-			arg0 := args[0]
-			if c, ok1 := arg0.(*expression.Column); ok1 {
-				col = c
+		case ast.Plus, ast.Minus, ast.Mul, ast.Div:
+			// partition by (col op const) where op in [+ - * /]
+			if arg1, ok1 := args[0].(*expression.Column); ok1 {
+				if _, ok2 := args[1].(*expression.Constant); ok2 {
+					if raw.FuncName.L == ast.Div {
+						monotonous = monotoneModeNonStrict
+					} else {
+						monotonous = monotoneModeStrict
+					}
+					return arg1, raw, monotonous, nil
+				}
+			}
```

Comment:
Why add those codes?  And there is no test for it?

Reply:
There are already test codes that cover the case.

In the beginning, this PR means to fix `partition by range (col1 + col2)`, but if we simply disable `col1 + col2` we will also disable `partition by range (col + const)`.

In the test code, some cases use `partition by range (a + 0)`, so this change is used to make those cases happy.

`col1 + col2`:  fail to do partition pruning
`col + const`:  can be pruned.

Reply:
How about:
```go
diff --git a/planner/core/rule_partition_processor.go b/planner/core/rule_partition_processor.go
index 107d2f29f..b80969eee 100644
--- a/planner/core/rule_partition_processor.go
+++ b/planner/core/rule_partition_processor.go
@@ -755,18 +755,6 @@ func makePartitionByFnCol(sctx sessionctx.Context, columns []*expression.Column,
                                        }
                                }
                        }
-               case ast.Plus, ast.Minus, ast.Mul, ast.Div:
-                       // partition by (col op const) where op in [+ - * /]
-                       if arg1, ok1 := args[0].(*expression.Column); ok1 {
-                               if _, ok2 := args[1].(*expression.Constant); ok2 {
-                                       if raw.FuncName.L == ast.Div {
-                                               monotonous = monotoneModeNonStrict
-                                       } else {
-                                               monotonous = monotoneModeStrict
-                                       }
-                                       return arg1, raw, monotonous, nil
-                               }
-                       }
                default:
                        fn = raw
                        monotonous = getMonotoneMode(raw.FuncName.L)
@@ -915,6 +903,10 @@ var monotoneIncFuncs = map[string]monotoneMode{
        ast.Year:          monotoneModeNonStrict,
        ast.ToDays:        monotoneModeNonStrict,
        ast.UnixTimestamp: monotoneModeStrict,
+       ast.Plus:          monotoneModeStrict,
+       ast.Minus:         monotoneModeStrict,
+       ast.Mul:           monotoneModeStrict,
+       ast.Div:           monotoneModeNonStrict,
 }

 func getMonotoneMode(fnName string) monotoneMode {
```

Reply:
Done